hanadb_exporter/main.py

```diff
@@ -139,39 +176,55 @@ def run():
     try:
         hana_config = config['hana']
         dbs = db_manager.DatabaseManager()
-        user=hana_config.get('user', '')
-        password=hana_config.get('password', '')
-        userkey=hana_config.get('userkey', None)
-        aws_secret_name=hana_config.get('aws_secret_name', '')
+        host = hana_config['host']
+        port = hana_config.get('port', 30013)
+        user = hana_config.get('user', '')
+        password = hana_config.get('password', '')
+        userkey = hana_config.get('userkey', None)
+        aws_secret_name = hana_config.get('aws_secret_name', '')
+        scale_out_mode = hana_config.get('scale_out_mode', False)
 
         if aws_secret_name:
             LOGGER.info('AWS secret name is going to be used to read the database username and password')
             db_credentials = secrets_manager.get_db_credentials(aws_secret_name)
             user = db_credentials["username"]
             password = db_credentials["password"]
 
-        dbs.start(
-            hana_config['host'], hana_config.get('port', 30013),
+        start(
+            db_manager=dbs, host=host, port=port,
             user=user,
             password=password,
             userkey=userkey,
+            scale_out_mode=scale_out_mode,
             multi_tenant=config.get('multi_tenant', True),
-            timeout=config.get('timeout', 30))
+            timeout=config.get('timeout', 30),
+            metrics=metrics)
     except KeyError as err:
         raise KeyError('Configuration file {} is malformed: {} not found'.format(args.config, err))
 
     if args.daemon:
         utils.systemd_ready()
 
-    connectors = dbs.get_connectors()
-    collector = prometheus_exporter.SapHanaCollectors(connectors=connectors, metrics_file=metrics)
-    REGISTRY.register(collector)
-    LOGGER.info('exporter successfully registered')
+    all_database_hosts = dbs.get_database_hosts()
 
     LOGGER.info('starting to serve metrics')
     start_http_server(config.get('exposition_port', 9668), '0.0.0.0')
     while True:
+        master_host = dbs.get_active_master_host(all_database_hosts, host, port)
+
+        if host != master_host:
+            host = master_host
+            start(
```

Comment:
In fact, we shouldn't have that many starts. We already started the manager before. Here you would register a new controller over again, which we must not do, as the collectors are registered at the beginning, before starting to server the http endpoint. I don't know which sort of side effect this could have

Reply:
I added comment at L:135, the connectors are registered once, I don't add the collector when scale-out is enabled on a non-master node, they will be add when a failover happen, if I add the collectors on a worker node they will export same data as master node and the main idea to keep the exporter as on standby mode on worker nodes.

Reply:
maybe the name start is confusing here